### PR TITLE
R4R: Refactor HTLC module

### PR DESCRIPTION
**Pull request by SegueII**
Resolves: #2095

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -202,21 +210,32 @@ func NewIrisApp(logger log.Logger, db dbm.DB, traceStore io.Writer, loadLatest b
...
+		slashing.ModuleName,
+		htlc.ModuleName,
+	)
 
-	app.mm.SetOrderEndBlockers(crisis.ModuleName, gov.ModuleName, staking.ModuleName)
+	app.mm.SetOrderEndBlockers(
+		crisis.ModuleName,
+		gov.ModuleName,
+		staking.ModuleName,
+		htlc.ModuleName,
```
Comment (chengwenxi) [inline app/app.go]: seems the HTLC endBlocker should be executed first

File: cli_test/htlc_test.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+// TODO: fix
+func TestIrisCLIHTLC(t *testing.T) {
+	t.Parallel()
+	fixtures := InitFixtures(t)
+
+	flags := fmt.Sprintf("--home=%s --node=%v --chain-id=%v --output=json", fixtures.IriscliHome, fixtures.RPCAddr, fixtures.ChainID)
+
+	// start iris server
+	proc := tests.GoExecuteTWithStdout(t, fmt.Sprintf("iris start --home=%s --rpc.laddr=%v --p2p.laddr=%v", fixtures.IrisdHome, fixtures.RPCAddr, fixtures.P2PAddr))
```
Comment (chengwenxi) [inline cli_test/htlc_test.go]: use` fixtures.start`, and replace `iriscli` by `fixtures.IriscliBinary`

File: config/config.go [hunk 3]
```diff
@@ -1,8 +1,10 @@
 package config
 
 const (
-	Testnet = "testnet"
-	Mainnet = "mainnet"
+	Testnet        = "testnet"
+	Mainnet        = "mainnet"
+	Iris           = "iris"
+	DefaultKeyPass = "1234567890"
```
Comment (chengwenxi) [inline config/config.go]: why defined here

File: modules/htlc/client/cli/query.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+	return mintingQueryCmd
+}
+
+// GetCmdQueryHTLC implements the query HTLC command.
+func GetCmdQueryHTLC(cdc *codec.Codec) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "query-htlc",
+		Short:   "Query details of an HTLC",
+		Example: "iriscli query htlc query-htlc <hash-lock>",
```
Comment (chengwenxi) [inline modules/htlc/client/cli/query.go]: `iriscli query htlc htlc`?

File: modules/htlc/client/cli/tx.go [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+			if len(secretStr) > 0 {
+				if len(secretStr) != 2*types.SecretLength {
+					return fmt.Errorf("the secret must be %d bytes long", types.SecretLength)
+				}
+
+				secret, err = hex.DecodeString(secretStr)
+				if err != nil {
+					return err
+				}
+			} else {
```
Comment (chengwenxi) [inline modules/htlc/client/cli/tx.go]: should update, see https://github.com/irisnet/irishub/pull/2110

File: modules/htlc/internal/types/events.go [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+	EventTypeCreateHTLC  = "create_htlc"
+	EventTypeClaimHTLC   = "claim_htlc"
+	EventTypeRefundHTLC  = "refund_htlc"
+	EventTypeExpiredHTLC = "expired_htlc"
+
+	AttributeValueSender               = "sender"
+	AttributeValueReceiver             = "receiver"
+	AttributeValueReceiverOnOtherChain = "receiver-on-other-chain"
+	AttributeValueAmount               = "amount"
+	AttributeValueHashLock             = "hash-lock"
```
Comment (chengwenxi) [inline modules/htlc/internal/types/events.go]: `-` -> `_`

File: modules/htlc/genesis_test.go [hunk 7]
```diff
@@ -1,100 +1,103 @@
...
+	newBlockHeight := int64(50)
+	suite.ctx = suite.ctx.WithBlockHeight(newBlockHeight)
+	modulehtlc.BeginBlocker(suite.ctx, suite.app.HTLCKeeper)
 
 	// export genesis
-	exportedGenesis := ExportGenesis(ctx, keeper)
+	exportedGenesis := modulehtlc.ExportGenesis(suite.ctx, suite.app.HTLCKeeper)
 	exportedHTLCs := exportedGenesis.PendingHTLCs
-	require.Equal(t, 1, len(exportedHTLCs))
+	require.Equal(suite.T(), 1, len(exportedHTLCs))
```
Comment (chengwenxi) [inline modules/htlc/genesis_test.go]: suite. Equal(), all `require` can be replaced with `suite`

File: modules/htlc/internal/keeper/keeper_test.go [hunk 8]
```diff
@@ -38,82 +69,80 @@ func TestKeeper_CreateHTLC(t *testing.T) {
 		state,
 	)
 
-	originSenderAccAmt := ak.GetAccount(ctx, senderAddr).GetCoins()
+	originSenderAccAmt := suite.app.AccountKeeper.GetAccount(suite.ctx, addrSender).GetCoins()
 
-	htlcAddr := auth.HTLCLockedCoinsAccAddr
-	require.Nil(t, ak.GetAccount(ctx, htlcAddr))
+	htlcAddr := suite.app.SupplyKeeper.GetModuleAddress(types.ModuleName)
+	require.Nil(suite.T(), suite.app.AccountKeeper.GetAccount(suite.ctx, htlcAddr))
```
Comment (chengwenxi) [inline modules/htlc/internal/keeper/keeper_test.go]: ditto.

File: modules/htlc/client/cli/query.go [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+	return mintingQueryCmd
+}
+
+// GetCmdQueryHTLC implements the query HTLC command.
+func GetCmdQueryHTLC(cdc *codec.Codec) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "htlc",
+		Short:   "Query details of an HTLC",
+		Example: "iriscli query htlc htlc <hash-lock>",
```
Comment (zhiqiang-bianjie) [inline modules/htlc/client/cli/query.go]: htlc htlc ?

File: modules/htlc/client/cli/tx.go [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+				hashLockStr := strings.TrimSpace(viper.GetString(FlagHashLock))
+				hashLock, err = hex.DecodeString(hashLockStr)
+				if err != nil {
+					return err
+				}
+			} else {
+				secretStr := strings.TrimSpace(viper.GetString(FlagSecret))
+				if len(secretStr) > 0 {
+					if len(secretStr) != 2*types.SecretLength {
+						return fmt.Errorf("the secret must be %d bytes long", types.SecretLength)
```
Comment (zhiqiang-bianjie) [inline modules/htlc/client/cli/tx.go]: "the secret must be %d bytes long" should be  "the length of the secret bytes must be %d"

File: modules/htlc/genesis.go [hunk 11]
```diff
@@ -24,7 +26,7 @@ func InitGenesis(ctx sdk.Context, k Keeper, data GenesisState) {
 func ExportGenesis(ctx sdk.Context, k Keeper) GenesisState {
 	pendingHTLCs := make(map[string]HTLC)
 
-	k.IterateHTLCs(ctx, func(hlock []byte, h HTLC) (stop bool) {
+	k.IterateHTLCs(ctx, func(hlock types.HTLCHashLock, h HTLC) (stop bool) {
 		if h.State == OPEN || h.State == EXPIRED {
```
Comment (zhiqiang-bianjie) [inline modules/htlc/genesis.go]:
```
 if h.State == OPEN {
   //
 }else if  h.State == EXPIRED {
  //
 }
```

File: modules/htlc/genesis_test.go [hunk 12]
```diff
@@ -1,100 +1,103 @@
...
-	accountKey := sdk.NewKVStoreKey("accountkey")
-	htlcKey := sdk.NewKVStoreKey("htlckey")
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
 
-	ms := store.NewCommitMultiStore(db)
-	ms.MountStoreWithDB(accountKey, sdk.StoreTypeIAVL, db)
-	ms.MountStoreWithDB(htlcKey, sdk.StoreTypeIAVL, db)
-	ms.LoadLatestVersion()
+	modulehtlc "github.com/irisnet/irishub/modules/htlc"
```
Comment (zhiqiang-bianjie) [inline modules/htlc/genesis_test.go]: modulehtlc -> htlc

File: modules/htlc/internal/keeper/keeper.go [hunk 13]
```diff
@@ -40,57 +46,47 @@ func (k Keeper) GetCdc() *codec.Codec {
...
-		return nil, err
+	// transfer the specified tokens to HTLC module address
+	if err := k.sk.SendCoinsFromAccountToModule(ctx, htlc.Sender, types.ModuleName, htlc.Amount); err != nil {
+		return err
 	}
 
-	// add to coinflow
-	ctx.CoinFlowTags().AppendCoinFlowTag(ctx, htlc.Sender.String(), auth.HTLCLockedCoinsAccAddr.String(), htlc.Amount.String(), sdk.CoinHTLCCreateFlow, "")
-
 	// set the HTLC
```
Comment (zhiqiang-bianjie) [inline modules/htlc/internal/keeper/keeper.go]: add Event for `SendCoinsFromAccountToModule`

File: modules/htlc/internal/keeper/keeper.go [hunk 14]
```diff
@@ -101,66 +97,50 @@ func (k Keeper) ClaimHTLC(ctx sdk.Context, hashLock []byte, secret []byte) (sdk.
 	// delete from the expiration queue
 	k.DeleteHTLCFromExpireQueue(ctx, htlc.ExpireHeight, hashLock)
 
-	// add to coinflow
-	ctx.CoinFlowTags().AppendCoinFlowTag(ctx, auth.HTLCLockedCoinsAccAddr.String(), htlc.To.String(), htlc.Amount.String(), sdk.CoinHTLCClaimFlow, "")
-
```
Comment (zhiqiang-bianjie) [inline modules/htlc/internal/keeper/keeper.go]: add Event for `SendCoinsFromModuleToAccount `

File: modules/htlc/internal/keeper/keeper.go (hunk 13 above)
Comment (chengwenxi) [reply modules/htlc/internal/keeper/keeper.go]: all bank sending logic has events.

File: modules/htlc/internal/keeper/keeper.go [hunk 15]
```diff
@@ -101,66 +97,50 @@ func (k Keeper) ClaimHTLC(ctx sdk.Context, hashLock []byte, secret []byte) (sdk.
...
 	// do the refund
-	if _, err := k.bk.SendCoins(ctx, auth.HTLCLockedCoinsAccAddr, htlc.Sender, htlc.Amount); err != nil {
-		return nil, err
+	if err := k.sk.SendCoinsFromModuleToAccount(ctx, types.ModuleName, htlc.Sender, htlc.Amount); err != nil {
+		return "", err
 	}
 
 	// update the state in HTLC
 	htlc.State = types.REFUNDED
 	k.SetHTLC(ctx, htlc, hashLock)
```
Comment (zhiqiang-bianjie) [inline modules/htlc/internal/keeper/keeper.go]: remove `htlc` ?

Comment (chengwenxi) [reply modules/htlc/internal/keeper/keeper.go]: good proposal. cc: @secret2830

File: modules/htlc/internal/types/genesis.go [hunk 16]
```diff
@@ -4,3 +4,8 @@ package types
 type GenesisState struct {
 	PendingHTLCs map[string]HTLC // claimable HTLCs
 }
+
+// ValidateGenesis checks if parameters are within valid ranges
+func ValidateGenesis(data GenesisState) error {
```
Comment (zhiqiang-bianjie) [inline modules/htlc/internal/types/genesis.go]: ?

File: simapp/app.go [hunk 17]
```diff
@@ -66,6 +69,7 @@ var (
 		staking.BondedPoolName:    {supply.Burner, supply.Staking},
 		staking.NotBondedPoolName: {supply.Burner, supply.Staking},
 		gov.ModuleName:            {supply.Burner},
+		htlc.ModuleName:           nil,
```
Comment (zhiqiang-bianjie) [inline simapp/app.go]: delete

File: modules/htlc/client/cli/tx.go [hunk 18]
```diff
@@ -84,7 +84,7 @@ func GetCmdCreateHTLC(cdc *codec.Codec) *cobra.Command {
 				secretStr := strings.TrimSpace(viper.GetString(FlagSecret))
 				if len(secretStr) > 0 {
 					if len(secretStr) != 2*types.SecretLength {
-						return fmt.Errorf("the secret must be %d bytes long", types.SecretLength)
+						return fmt.Errorf("length of the secret bytes must be %d", types.SecretLength)
```
Comment (chengwenxi) [inline modules/htlc/client/cli/tx.go]: should be `2*types.SecretLength`

Comment (SegueII) [reply modules/htlc/client/cli/tx.go]: length of "secret bytes" should be "types.SecretLength" here
should be change to "2*types.SecretLength" if using "secret str"

Comment (chengwenxi) [reply modules/htlc/client/cli/tx.go]: using secret str? I think it's better.

File: modules/htlc/internal/keeper/keeper.go (hunk 15 above)
Comment (SegueII) [reply modules/htlc/internal/keeper/keeper.go]: Delete the refunded HTLC #2128

File: modules/htlc/client/rest/rest.go [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+
+// RegisterRoutes registers HTLC-related REST handlers to a router
+func RegisterRoutes(cliCtx context.CLIContext, r *mux.Router) {
+	registerQueryRoutes(cliCtx, r)
+	registerTxRoutes(cliCtx, r)
+}
+
+// CreateHTLCReq defines the properties of a create HTLC request's body.
+type CreateHTLCReq struct {
+	BaseTx               rest.BaseReq   `json:"base_tx" yaml:"base_tx"`
```
Comment (secret2830) [inline modules/htlc/client/rest/rest.go]: Maybe BaseReq is better?

File: simapp/app.go [hunk 20]
```diff
@@ -237,21 +244,32 @@ func NewSimApp(
...
-	app.mm.SetOrderBeginBlockers(mint.ModuleName, distr.ModuleName, slashing.ModuleName)
-	app.mm.SetOrderEndBlockers(crisis.ModuleName, gov.ModuleName, staking.ModuleName)
+	app.mm.SetOrderBeginBlockers(
+		mint.ModuleName,
+		distr.ModuleName,
+		slashing.ModuleName,
+		htlc.ModuleName,
+	)
+	app.mm.SetOrderEndBlockers(
+		htlc.ModuleName,
```
Comment (secret2830) [inline simapp/app.go]: No EndBlocker in HTLC

**Pull request merged**